scene/resources/tile_set.cpp

```diff
@@ -4797,24 +4785,86 @@ void TileSetAtlasSource::_update_padded_texture() {
-			image->set_pixelv(base_pos + Vector2i(src_rect.size.x, src_rect.size.y), src->get_pixelv(src_rect.position + Vector2i(src_rect.size.x - 1, src_rect.size.y - 1)));
+			image->set_pixelv(base_pos + Vector2i(-1, -1), src_image->get_pixelv(src_rect.position));
+			image->set_pixelv(base_pos + Vector2i(src_rect.size.x, -1), src_image->get_pixelv(src_rect.position + Vector2i(src_rect.size.x - 1, 0)));
+			image->set_pixelv(base_pos + Vector2i(-1, src_rect.size.y), src_image->get_pixelv(src_rect.position + Vector2i(0, src_rect.size.y - 1)));
+			image->set_pixelv(base_pos + Vector2i(src_rect.size.x, src_rect.size.y), src_image->get_pixelv(src_rect.position + Vector2i(src_rect.size.x - 1, src_rect.size.y - 1)));
 		}
 	}
 
-	if (!padded_texture.is_valid()) {
-		padded_texture.instantiate();
+	Ref<ImageTexture> image_texture;
+	image_texture.instantiate();
+	image_texture->set_image(image);
+
+	return image_texture;
+}
+
+void TileSetAtlasSource::_update_padded_texture() {
+	if (!padded_texture_needs_update) {
+		return;
+	}
+	padded_texture_needs_update = false;
+	padded_texture.instantiate();
+
+	if (!texture.is_valid()) {
+		return;
+	}
+
+	if (!use_texture_padding) {
+		return;
+	}
+
+	CanvasTexture *src_canvas_texture = Object::cast_to<CanvasTexture>(texture.ptr());
+	if (src_canvas_texture) {
+		// Use all textures
+		// Diffuse
+		Ref<Texture2D> src = src_canvas_texture->get_diffuse_texture();
+		Ref<ImageTexture> image_texture;
+		if (src.is_valid()) {
+			image_texture = _create_padded_image_texture(src);
+		} else {
+			image_texture.instantiate();
+		}
+		padded_texture->set_diffuse_texture(image_texture);
+
+		// Normal
+		src = src_canvas_texture->get_normal_texture();
+		image_texture.instantiate();
+		if (src.is_valid()) {
+			image_texture = _create_padded_image_texture(src);
+		} else {
+			image_texture.instantiate();
+		}
+		padded_texture->set_normal_texture(image_texture);
+
+		// Specular
+		src = src_canvas_texture->get_specular_texture();
+		image_texture.instantiate();
+		if (src.is_valid()) {
+			image_texture = _create_padded_image_texture(src);
+		} else {
+			image_texture.instantiate();
+		}
+		padded_texture->set_specular_texture(image_texture);
+
+		// Other properties
+		padded_texture->set_specular_color(src_canvas_texture->get_specular_color());
+		padded_texture->set_specular_shininess(src_canvas_texture->get_specular_shininess());
+		padded_texture->set_texture_filter(src_canvas_texture->get_texture_filter());
+		padded_texture->set_texture_repeat(src_canvas_texture->get_texture_repeat());
+	} else {
+		// Use only diffuse.
+		Ref<ImageTexture> image_texture = _create_padded_image_texture(texture);
+		padded_texture->set_diffuse_texture(image_texture);
 	}
-	padded_texture->set_image(image);
+	padded_texture->connect(SNAME("changed"), callable_mp(this, &TileSetAtlasSource::_queue_update_padded_texture));
```

Comment:
You can use Resource::connect_changed` here. Same for disconnect if you have any.